renku/core/management/migrations/m_0009__new_metadata_storage.py

```diff
@@ -0,0 +1,295 @@
+from renku.core.models.dataset import DatasetsProvenance
+from renku.core.models.jsonld import load_yaml
+from renku.core.models.project import Project
+from renku.core.models.provenance.activities import Activity
+from renku.core.models.provenance.activity import ActivityCollection
+from renku.core.models.workflow.dependency_graph import DependencyGraph
+from renku.core.utils import communication
+from renku.core.utils.migrate import MigrationType, read_project_version_from_yaml
+from renku.core.utils.scm import git_unicode_unescape
+
+GRAPH_METADATA_PATHS = [
+    Path(RENKU_HOME) / RepositoryApiMixin.DATABASE_PATH,
+    Path(RENKU_HOME) / RepositoryApiMixin.DEPENDENCY_GRAPH,
+    Path(RENKU_HOME) / RepositoryApiMixin.PROVENANCE_GRAPH,
+    Path(RENKU_HOME) / DatasetsApiMixin.DATASETS_PROVENANCE,
+]
+
+
+def migrate(client):
+    """Migration function."""
+    # NOTE: Don't run this migration for dataset-only migrations
+    if MigrationType.WORKFLOWS not in client.migration_type:
+        return
+
+    # TODO: set remove=True once the migration to the new metadata is finalized
+    generate_new_metadata(client, remove=False)
+    _migrate_project(client)
+
+
+def generate_new_metadata(client: LocalClient, force=True, remove=True):
+    """Generate graph and dataset provenance metadata."""
+    if force:
+        client.remove_graph_files()
+    elif client.has_graph_files():
+        raise errors.OperationError("Graph metadata exists.")
+
+    database = Database.from_path(client.database_path)
+    initialize_database(database)
+
+    dependency_graph = DependencyGraph.from_database(database)
+    datasets_provenance = DatasetsProvenance(database)
+
+    commits = list(client.repo.iter_commits(paths=[f"{client.workflow_path}/*.yaml", ".renku/datasets/*/*.yml"]))
+    n_commits = len(commits)
+    commits = reversed(commits)
+
+    for n, commit in enumerate(commits, start=1):
+        communication.echo(f"Processing commits {n}/{n_commits} {commit.hexsha}", end="\n")
+
+        try:
+            _process_workflows(
+                client=client, commit=commit, database=database, dependency_graph=dependency_graph, remove=remove
+            )
+            _process_datasets(client=client, commit=commit, datasets_provenance=datasets_provenance)
+        except errors.MigrationError:
+            communication.echo("")
+            communication.warn(f"Cannot process commit '{commit.hexsha}' - Migration failed: {traceback.format_exc()}")
+        except Exception:
+            communication.echo("")
+            communication.warn(f"Cannot process commit '{commit.hexsha}' - Exception: {traceback.format_exc()}")
+
+        # NOTE: We need to commit in case we want to resume because we delete processed files
+        database.commit()
+
+
+def _process_workflows(
+    client: LocalClient, commit: Commit, database: Database, dependency_graph: DependencyGraph, remove: bool
+):
+    def replace_agents(activity):
```

Comment:
A docstring describing why this function exists/what its use is would be nice. I don't think it's immediately apparent, especially to someone new.